src/solver/solver.cxx

```diff
@@ -468,8 +468,28 @@ int Solver::solve(int NOUT, BoutReal TIMESTEP) {
     throw BoutException(_("Failed to initialise solver-> Aborting\n"));
   }
 
+  // Set the run ID to a random number
+  run_restart_from = run_id; // Restarting from the previous run ID
```

Comment:
It would be nice to have `run_restart_from` be a `std::vector<unsigned_int>`, so if you restart from a restart (etc.) you know the IDs of all the previous runs. I think this needs an extra method or two adding to `DataFile` to handle `std::vector` (or `Array`?), but shouldn't be too complicated because `DataFormat` already supports C-arrays.

Reply:
This would be nice! These might need to be vectors in dump files, as at the moment if a simulation restarts and appends to the .dmp files then the run_id and run_restart_from will be overwritten. In that case possibly `run_id` should be a vector of the runs whose outputs are contained in the dump file, and `run_restart_from` could be a vector of previous runs?

Reply:
I'm having a go at adding `std::vector<int>` to `Datafile`s. Hopefully PR coming soon...

Reply:
See #2150, which allows a `std::vector<int>` to be added to a `Datafile`.

Not sure what to do about simulations that append. In #2150 I've tried to treat a `std::vector<int>` generically - I named the dimension "vec<n>" so we could add any number of vectors of any size, but that doesn't make sense if we want to append to them.

It wouldn't be crazy to have a dimension that we append one entry per run to - I'd actually quite like to have that for saving the git hashes of BOUT++ and physics models to - but that would be a bigger change to the `Datafile` interface. I hesitate to put more work into `Datafile` as we're going to (?) replace it with `OptionsNetCDF` in v5. One possible temporary work-around would be to just make `run_id` a time-dependent `int` variable. It's only a slight waste of disk space and would cure the appending problem. I should think we can put a special case into `xbout`'s `create_restarts()` method to get rid of duplicate entries and turn the time-dependent `run_id` into a `run_restart_from` vector. Then I guess we skip writing `run_restart_from` to dump files if we're appending, as the info would already be there in the original `run_restart_from` (if there was one) plus time-dependent `run_id`. Finally, open a feature request for whatever I/O system we end up with in v5 to tidy this up.

Reply:
I've started work on replacing `Datafile` with `OptionsNetCDF`, although not very far in yet.

Having a "number of restarts-length" unlimited dimension is a good idea. In the `OptionsNetCDF` approach, we also add everything in `Options` to the output file. This includes all the input parameters. Given that these could change in a restarted simulation, we would have to make basically _everything_ have this dimension.

An alternative approach might be to create a group structure:

```
/
|-- time evolving variables
| |-- n
| \-- T
|-- build info
| |-- run001
| | |-- git hash
| | |-- has_petsc
| | ...
| \-- run002
|   |-- git hash
|   |-- has_petsc
|   ...
\-- input parameters
  |-- run001
  | |-- run_id
  | |-- nx
  | ...
  \-- run002
    |-- run_id
    |-- nx
    ...
```

or perhaps

```
/
|-- time evolving variables
| |-- n
| \-- T
|-- run001
| |-- run_id
| |-- build info
| | |-- git hash
| | |-- has_petsc
| | ...
| \-- input parameters
|   |-- nx
|   ...
|-- run001
| |-- run_id
| |-- build info
| | |-- git hash
| | |-- has_petsc
| | ...
| \-- input parameters
|   |-- nx
|   ...
...
```

I know that xarray `Dataset` doesn't understand groups, but if we know what groups we're expecting `xBOUT`  can handle it -- flattening the separate run groups for instance.

Reply:
I quite like this group structure approach: If the evolving variables are in the root directory as now, and the metadata is in groups, then it should be backwards compatible. Then we're not having to add lots of dimensions to things that probably don't need them most of the time.

Reply:
This sounds like a pretty nice idea. Would need an xbout update, but seems like it should be straightforward and would make things nicer. At the moment xbout puts all scalars into a 'metadata' attribute, because most of them are settings (like `nx`, etc.), but there could also be scalar variables which it would be better to keep separate. xarray can load groups from a netCDF file (using the `group` kwarg to `open_dataset`), so we could instead create the metadata with something like
```
ds.attrs["metadata"] = open_dataset(\<first_file_paths\>, group="run001/input parameters")
```
or something a bit fancier to deal with multiple runs that might have the same input parameters (so we can use a single `metadata`) or might have changed values (concatenate the Datasets along a 'run' dimension maybe?). Anyway seems like a nice thing to have.

While we're discussing - it might be nice to have the non-time-dependent variables (`SAVE_ONCE()` ones) in the `run001`, `run002`, ..., groups in case they change. I'm not sure if we overwrite or keep the original value when appending at the moment...

Reply:
All that said, I would love a backportable version of this PR (i.e. works with existing `Datafile`), even if (for example) it doesn't support using `append` nicely.